### ISSUE TITLE
Problems with some LRO NAC datasets

### ISSUE THREAD

**Issue by oleg-alexandrov** [USGS-Astrogeology/usgscsm#357]
I have 75 LRO NAC datasets very close to the South Pole (within 0.2 degrees or so). I made CSM .json files for all of them, and evaluated the accuracy of projecting rays from the ISIS model onto the ground and back to the CSM model and vice versa. For 56 of them the pixel-to-pixel error is under 0.5 pixels. For some, it is 100 pixels, for some it becomes 70,000 pixels.

I must say that I am still using ISIS 5.0.1, ale 0.8.5, and ISISDATA likely a year old, so there is a chance the issue is on my side. I am surprised however, that for most datasets things work out, but for some they don't. 

I am attaching for the record the problematic .cub files, in case somebody has any quick ideas. 

The list below has the  image name and pixel-to-pixel error, and for completeness listed are a handful of datasets for which this is error actually very good. The worst are towards the bottom.

M169576603LE    0.233772
M139831449LE    0.234603
M139831449RE    0.234668
M139729686LE    0.235561
M141059693LE    0.242376
M169488450LE    0.245601
M169474886LE    0.246625
M157830228LE    0.257869
M169468103LE    0.261228
M157830228RE    0.261858
M169502013RE    0.264729
M169556262RE    0.287273
M169549481RE    0.300138
M169549481LE    0.302653
M157809893RE    0.307175
M157843789LE    0.30771
M1315066257LE   0.34057
M169278138LE    0.354418
M169291703LE    0.369411
M139716114RE    0.397297
M169522357LE    0.434125
M169508794LE    0.437499
M150366876RE    3.12818
M148012947LE    3.6963
M110962901RE    7.94198
M113679075LE    15.034
M147992619RE    25.0667
M108946135RE    34.354
M167289893LE    353.674
M121014786LE    747.538
M162587237LE    926.804
M162580453LE    963.458
M121028353LE    975.651
M162553316RE    1233.43
M162499044LE    1593.95
M162492261LE    1858.43
M162485477LE    2138.01
M162478694LE    2404.19
M114859732RE    51393.9
M119582918RE    61134.2
M117241732LE    70079.7

**Comment by jessemapel** [comment]
Is any of this limb data? CSM and ISIS handle the limb quite differently. CSM will give you the closest point on the body to the look vector while ISIS will just return a non-intersection.

I'd also double check your surface model and make sure you're using the same surface model intersection routine.

**Comment by jessemapel** [comment]
For surface model and surface model intersection, I'm talking about what ground elevation data are you using to do this step? An ellipsoid? A DEM? Are you using the same routines to compute the intersections? CSM only supports very naive bi-axial ellipsoid intersections in its image2ground method. If you want to intersect a better surface model, like a DEM, then you need to have some extra code.

> projecting rays from the ISIS model onto the ground and back to the CSM model and vice versa.

For limb data, I mean imagery that views the limb of the body, highly oblique imagery.

**Comment by oleg-alexandrov** [comment]
I did spend some time this morning fiddling with knoten and I was not getting useful answers even for a cub I knew to be good. 

At some point may be nice to have a single C++ tool in the ISIS repo which takes cameras as input and prints a report for CSM vs ISIS, rather than the existing various layers of, conda, Python and notebooks. 

I guess there's no quick solution here. Anyhow, hopefully the list of cubs I provide above may be of some use if at some point somebody may have time for quality control.

**Comment by oleg-alexandrov** [comment]
I spent a little more time on this. The issue seems to be in in the linescan sensor implementation of usgscsm. The image to ground function seems fine, it produces a point on the ground which then the ISIS model projects back correctly in the image.

Yet, in the ground to image function of usgscsm, sensible ground points (obtained by intersecting with the ground a ray starting at the center of the image sensor), refuse to project back into the image with the CSM model. The logic we added here:

 https://github.com/USGS-Astrogeology/usgscsm/blob/dev/src/UsgsAstroLsSensorModel.cpp#L652

fails to work with either positive or negative m_iTransL. 

The obliqueness does not seem to be an issue. The ray from camera to ground and ray from camera to planet center make an angle of at most 3.2 degrees. This is for problematic dataset M167289893LE (one of the ones mentioned above). 

For comparison, a dataset that performs well, which I used for testing last year, named M1127782730LE, as an off-nadir angle of 12 degrees yet the ground to image logic converges. 

In short, seems like a nasty thing somewhere in the algorithm. 

I verified that this fails with the latest code from GitHub. 

It could still be a polar business issue, but not related to obliqueness.

**Comment by oleg-alexandrov** [comment]
I took another look at this and apparently the problem was simple, the algorithm was overshooting. I put a fix and made a pull request at https://github.com/USGS-Astrogeology/usgscsm/pull/358.

**Comment by jessemapel** [comment]
I've done some more looking into this and I'm not sure if it's just the algorithm overshooting. This seems to be more of the algorithm just assuming things are nicely linear when they are actually quite non-linear. Here's a gist of some of my testing:

https://gist.github.com/jessemapel/7998c8361575c17f769710ccb3c34782

If you look at the graphs which show the groundtoimage attempting to converge at the worst test point in each image, you can clearly see it diverging as it gets close to the correct solution. The red dashed lines are the image boundaries and the green line is the actual line it should be converging to.

I'm going to re-run the tests today and revert back to the older algorithm we ported over from ISIS. It is slower, but seems to handle these situations better.

The other issue you can clearly see is the initial linear approximations is not very high quality. Maybe something like a piecewise linear approximation or a more complex approximation mapping would help save some time lost to the slower algorithm.

**Comment by oleg-alexandrov** [comment]
Thank you for looking at this. This is good to know.

The situation can be quite nonlinear, I guess, if, for example, the camera goes South but it pans North as it does so, and its trajectory is not quite a gentle arc. 

I am rather confident that, as you say, an initial guess can be found with a nonlinear approximation or interpolation into some lookup table of moderate size, as the sensor itself is underpinned by a lookup table as well. (I could look into this myself, later this year, maybe.)

In the meantime I've been stress-testing this current logic with more datasets from the South pole. The step size halving I did in the pull request also fails sometimes, and making the step size 1/1000th of the current one fails as well, and in fact it can blow up, as you see too. My solution for now is to start (in the constructor) with a sample of possible steps, from current one, 1/2, 1/4, ... of it, and see which results in best convergence, and stick with that for the whole sensor. 

In the meantime, I must say we are using CSM instead of the ISIS sensor for shape-from-shading, where an outrageous amount of ground to image computations happen per each single pixel as it iterates, for lots of very big images, and it is much, much faster than ISIS, is great, as that was a very serious bottleneck.

**Comment by jessemapel** [comment]
For the initial approximation something like this seems like a good candidate:

https://khanhha.github.io/posts/Thin-Plate-Splines-Warping/

This example talks about warping of image coordinates (so 2D -> 2D), but the math should also work for mapping object coordinates into image coordinates (3D -> 2D). It'll require a bit more computation at the start up because it requires some extra spline fitting, but it shouldn't be too costly as it's a purely linear problem (with respect to the spline coefficients).

I'm going to think about this and maybe tinker with an implementation

**Comment by oleg-alexandrov** [comment]
Fair enough. I am more in favor of an RPC fit (sorry to always bring it up), as that is the default approach for Earth satellites (DigitalGlobe, Pléiades, Planet, CartoSat, PeruSat) when doing ground to image. They don't even bother to do iterations. But that can be tricky to fit. One can start with the projective transform (ratio of polynomials of degree 1) which is how one does ground-to-image for pinhole cameras, and do further refinements. 

But if you can find thin-plate ready made implementation, sure. With that, one could actually avoid an iterative algorithm altogether, but the table you may need to store may need to have samples in 3 dimensions, whose number can quickly grow.

**Comment by jessemapel** [comment]
True, an initial RPC fit is much more in line with best practices

**Comment by oleg-alexandrov** [comment]
My experience with it is that RPC may be good enough for initial guess but iterations may be needed afterwards like now. I should not try to sway you one way or another as going down either path will be some work. 

All I can say is that both of these methods will need a validity box, you can't fit them or tabulate them in whole 3D space. A natural domain is to look at is the "pyramid" of the camera view. (I know the camera is linescan, but one can think of it as pinhole for a little). So, its tip is the camera center, and its "base" is generally in the direction of the planet. One can choose two such "bases", at various heights, covering whatever possible ground points one may ever need, and do a fit in there.

This will not be little work. I think discussing this is fun, but I would not be sure about the strategy or timing for such work.

**Comment by jessemapel** [comment]
I agree this is non-trivial. We have funding to work this fiscal year on the project but it is already assigned to other tasks. I may be able to make it fit if we complete the existing tasks under estimate. If it doesn't fit into this fiscal year's work, I will propose for it under next fiscal year's work.

**Comment by jessemapel** [comment]
Fixed by #358 I'll open a new issue for a better initial approximation.
